src/commands/to_release.rs

```diff
@@ -103,17 +126,293 @@ where
         }
     }
 
-    let indices = petgraph::algo::toposort(&graph, None).map_err(|c| {
-        format!(
-            "Cycle detected: {:}",
-            graph.node_weight(c.node_id()).unwrap().name()
-        )
-    })?;
-    let packages = indices
+    if let Some(dest) = write_dot_graph {
+        let mut dest = OpenOptions::new()
+            .create(true)
+            .truncate(true)
+            .write(true)
+            .open(dest)?;
+        graphviz(&graph, &mut dest)?;
+    }
+
+    // cannot use `toposort` for graphs that are cyclic in a undirected sense
+    // but are not in a directed way
+    let mut cycles = vec![];
+    let mut toposorted_indices = vec![];
+    let strongly_connected_sets = petgraph::algo::kosaraju_scc(&graph);
+    for strongly_connected in strongly_connected_sets {
+        match strongly_connected.len() {
+            0 => unreachable!("Strongly connected components are at least size 1. qed"),
+            1 => toposorted_indices.push(strongly_connected[0]),
+            _ => cycles.push(strongly_connected),
+        }
+    }
+    if !cycles.is_empty() {
+        assert!(petgraph::algo::is_cyclic_directed(&graph));
+        let cycles = cycles
+            .iter()
+            .map(|x| {
+                x.iter()
+                    .map(|i| graph.node_weight(*i).unwrap())
+                    .map(|pkg| pkg.name())
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>();
+        anyhow::bail!("Contains cycles: {:?}", cycles);
+    }
+
+    // reverse in place, the output of `scc_karaju` is in reverse order
+    toposorted_indices.reverse();
+
+    let packages = toposorted_indices
         .into_iter()
         .map(|i| graph.node_weight(i).unwrap().clone())
         .rev()
         .collect::<Vec<_>>();
 
     Ok(packages)
 }
+
+/// Render a graphviz (aka dot graph) to a file.
+fn graphviz(graph: &Graph<Package, ()>, dest: &mut impl Write) -> anyhow::Result<()> {
+    let config = &[dot::Config::EdgeNoLabel, dot::Config::NodeNoLabel][..];
+    let dot = Dot::with_attr_getters(
+        graph,
+        config,
+        &|_graph, _edge_ref| -> String { "".to_owned() },
+        &|_graph, (_idx, pkg)| -> String {
+            format!(
+                r#"label="{}:{}""#,
+                pkg.name().to_string(),
+                pkg.version().to_string().as_str()
+            )
+        },
+    );
+    dest.write_all(format!("{:?}", &dot).as_bytes())?;
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
we usually use actual fixtures and cli-e2e tests for cases this complicated, but this is fine, too...!